session/nontransactional.go

```diff
@@ -0,0 +1,509 @@
+	if !(se.GetSessionVars().IsAutocommit() && !se.GetSessionVars().InTxn()) {
+		return nil, errors.Errorf("non-transactional statement can only run in auto-commit mode. auto=commit:%v, inTxn:%v",
+			se.GetSessionVars().IsAutocommit(), se.GetSessionVars().InTxn())
+	}
+	tableName, selectSQL, shardColumn, err := buildSelectSQL(stmt, se)
+	if err != nil {
+		return nil, err
+	}
+	if stmt.DryRun == ast.DryRunQuery {
+		return buildDryRunResults(stmt.DryRun, []string{selectSQL}, se.GetSessionVars().BatchSize.MaxChunkSize)
+	}
+	jobs, err := buildShardJobs(ctx, stmt, se, selectSQL, shardColumn)
+	if err != nil {
+		return nil, err
+	}
+
+	splitStmts, err := splitDeleteWorker(ctx, jobs, stmt, tableName, se, stmt.DeleteStmt.Where)
+	if err != nil {
+		return nil, err
+	}
+	if stmt.DryRun == ast.DryRunSplitDml {
+		return buildDryRunResults(stmt.DryRun, splitStmts, se.GetSessionVars().BatchSize.MaxChunkSize)
+	}
+	return buildExecuteResults(jobs, se.GetSessionVars().BatchSize.MaxChunkSize)
+}
+
+// single-threaded worker. work on the key range [start, end]
+func splitDeleteWorker(ctx context.Context, jobs []job, stmt *ast.NonTransactionalDeleteStmt,
+	tableName *ast.TableName, se Session, originalCondition ast.ExprNode) ([]string, error) {
+
+	// prepare for the construction of statement
+	var shardColumnRefer *ast.ResultField
+	var shardColumnType types.FieldType
+	for _, col := range tableName.TableInfo.Columns {
+		if col.Name.L == stmt.ShardColumn.Name.L {
+			shardColumnRefer = &ast.ResultField{
+				Column: col,
+				Table:  tableName.TableInfo,
+			}
+			shardColumnType = col.FieldType
+		}
+	}
+	if shardColumnRefer == nil && stmt.ShardColumn.Name.O != "_tidb_rowid" {
+		return nil, errors.New("Non-transactional delete, column not found")
+	}
+
+	splitStmts := make([]string, 0, len(jobs))
+	for i := range jobs {
+		select {
+		case <-ctx.Done():
+			failedJobs := make([]string, 0)
+			for _, job := range jobs {
+				if job.err != nil {
+					failedJobs = append(failedJobs, fmt.Sprintf("job:%s, error: %s", job.String(), job.err.Error()))
+				}
+			}
+			if len(failedJobs) == 0 {
+				logutil.Logger(ctx).Warn("Non-transactional delete worker exit because context canceled. No errors",
+					zap.Int("finished", i), zap.Int("total", len(jobs)))
+			} else {
+				logutil.Logger(ctx).Warn("Non-transactional delete worker exit because context canceled. Errors found",
+					zap.Int("finished", i), zap.Int("total", len(jobs)), zap.Strings("errors found", failedJobs))
+			}
+			return nil, ctx.Err()
+		default:
+		}
+
+		stmtBuildInfo := statementBuildInfo{
+			stmt:              stmt,
+			shardColumnType:   shardColumnType,
+			shardColumnRefer:  shardColumnRefer,
+			originalCondition: originalCondition,
+		}
+		splitStmt := doOneJob(ctx, &jobs[i], len(jobs), stmtBuildInfo, se, stmt.DryRun == ast.DryRunSplitDml)
+		splitStmts = append(splitStmts, splitStmt)
+		// if the first job failed, there is a large chance that all jobs will fail. So return early.
```

Comment:
In this case, no delete is actually applied, is it right?

Reply:
Yes.

Reply:
What about response error with this information?